internal/wrphandlers/qos/priority_queue.go

```diff
@@ -0,0 +1,99 @@
+// SPDX-FileCopyrightText: 2024 Comcast Cable Communications Management, LLC
+// SPDX-License-Identifier: Apache-2.0
+
+package qos
+
+import (
+	"container/heap"
+
+	"github.com/xmidt-org/wrp-go/v3"
+)
+
+// priorityQueue implements heap.Interface and holds wrp Message, using wrp.QOSValue as its priority.
+// https://xmidt.io/docs/wrp/basics/#qos-description-qos
+type priorityQueue struct {
+	// queue for wrp messages, ingested by serviceQOS
+	queue []wrp.Message
+	// maxQueueSize is the allowable max size of the queue based on the sum of all queued wrp message's payloads
+	maxQueueSize int
+}
+
+// Dequeue returns the next highest priority message.
+func (pq *priorityQueue) Dequeue() (wrp.Message, bool) {
+	// Required, otherwise heap.Pop will panic during an internal Swap call.
+	if pq.Len() == 0 {
+		return wrp.Message{}, false
+	}
+
+	msg, ok := heap.Pop(pq).(wrp.Message)
+
+	return msg, ok
+}
+
+// Enqueue queues the given message.
+func (pq *priorityQueue) Enqueue(msg wrp.Message) {
+	// Check whether msg would already violate maxQueueSize.
+	if len(msg.Payload) > pq.maxQueueSize {
+		return
+	}
+
+	// Check whether enqueuing msg would violate maxQueueSize.
```

Comment:
I'd add the new message, then create a new function:
```golang
func (pq *priorityQueue) drop() {
    _ = heap.Remove(pq, pq.Len()-1)
}
```

This makes the code in the for loop much simpler to read and understand because you'll just loop until the size is less than or equal to the max size.

Reply:
agree 🙂 

From our side conversation it sounded like we're considering making this a configurable feature, right?
I.e. `should we favor old or new messages [when the queue is full]?`

- If `we favor old messages`, then we want to check whether the oldest message is of higher priority than the new message - keeping the old and dropping the new message.
- If `we favor new messages`, then we always enqueue the new message while dropping the oldest message.